src/uct/ib/rc/accel/rc_mlx5_common.c

```diff
@@ -346,41 +349,41 @@ static void uct_mlx5_iface_cleanup_dm_tl(uct_mlx5_dm_data_t *data)
 }
 #endif
 
-static ucs_status_t uct_mlx5_iface_init_dm(uct_rc_mlx5_iface_common_t *iface,
-                                           uct_rc_iface_t *rc_iface,
-                                           const uct_common_mlx5_iface_config_t *config)
+static ucs_status_t
+uct_rc_mlx5_iface_common_dm_init(uct_rc_mlx5_iface_common_t *iface,
+                                 uct_rc_iface_t *rc_iface,
+                                 const uct_common_mlx5_iface_config_t *config)
 {
 #if HAVE_IBV_EXP_DM
     if ((config->dm.seg_len * config->dm.count) == 0) {
-        iface->dm.dm = NULL;
-        return UCS_OK;
+        goto fallback;
     }
 
     iface->dm.dm = uct_worker_tl_data_get(rc_iface->super.super.worker,
                                           UCT_IB_MLX5_WORKER_DM_KEY,
                                           uct_mlx5_dm_data_t,
-                                          uct_mlx5_iface_dm_device_cmp,
-                                          uct_mlx5_iface_init_dm_tl, rc_iface, config);
-    if (iface->dm.dm == NULL) {
-        return UCS_OK;
-    }
+                                          uct_rc_mlx5_iface_common_dm_device_cmp,
+                                          uct_rc_mlx5_iface_common_dm_tl_init,
+                                          rc_iface, config);
+    if (UCS_PTR_IS_ERR(iface->dm.dm)) {
+        goto fallback;
+    }
+    ucs_assert(iface->dm.dm->dm != NULL);
+    iface->dm.seg_len = iface->dm.dm->seg_len;
+    return UCS_OK;
 
-    if (iface->dm.dm->dm == NULL) {
-        /* failed to initialize DM. put back object */
-        uct_worker_tl_data_put(iface->dm.dm, uct_mlx5_iface_cleanup_dm_tl);
-        iface->dm.dm = NULL;
-    } else {
-        iface->dm.seg_len = iface->dm.dm->seg_len;
-    }
+fallback:
+    iface->dm.dm      = NULL;
+    iface->dm.seg_len = UCT_IB_MLX5_AM_MAX_SHORT(0);
```

Comment:
better not overload iface->dm.seg_len to hold max_inline, but check if dm is present in query_iface and return either dm seg len or BF reg size